### PR TITLE
LF Script providing function to doing the post processing

### PR DESCRIPTION
**Pull request by thk123**
The function allows specifying input and output directories.

### CODE REVIEW

**Comment by igorclark** [comment]
Thanks a lot Thomas, from the sample output files, this seems to work great for a specific IU/scenario combination, but I'm having trouble getting it to work on the actual LF output. Also it's single threaded so it takes a very long time for thousands of results.

1. Here's some test data that I can reproduce the problem with, taken directly from the actual LF result data set.

	- Inputs, in a directory hierarchy exactly like the result data:
		https://storage.googleapis.com/ntd-disease-simulator-data/endgame-exports/20240905/lf-post-processing-5-samples-AGO.tbz

	- Outputs, as generated by the script:
		https://storage.googleapis.com/ntd-disease-simulator-data/endgame-exports/20240905/lf-post-processing-output-5-samples-AGO.tbz

	The intermediate and aggregate post-processing output files are all empty apart from column headers and some basic info.

	Any ideas on what's happening here would be great, I suspect it's perhaps because the LF data files are a slightly different format from the Oncho ones?

2. As it's single-threaded and we've got 10248 of those to churn through, it takes a veeeerry long time to work through them, and doesn't take any advantage of extra cores.

	(The reason we don't "feel" this length-of-time pain when we already do this same post-processing in Oncho is that it's happening as an automatic post-process step during the runs, so (a) it's happening ~128 at a time, and (b) the time is absorbed into the overall oncho-IU/scenario-combo time that's already veeery long.)

	So ... I wonder if it'll be possible to split into 2 scripts? The first one would take 2 arguments, 1 path to 1 CSV file and 1 output directory, and write out the `post_processed` file there, and then the second would do the aggregation of the `post_processed` files. That way we'd get 2 benefits - firstly we could use it as an integrated per-IU post-processing step as part of future runs, and secondly I can do the intermediate step on the existing data 128 times faster than now, and then we can always do the aggregation as a single step, like in Oncho. (We know from Oncho that the latter step is fast, even single-threaded.)

	(As a sort-of-side-note, though: for SCH/STH and Trachoma, which do multi-processing, we'd want this also to use multiple cores in Python, as due to the multi-core nature of those models, the runners run them on IUs sequentially, so a single-threaded post-processing step would mean 127 cores sitting idle while waiting for 1 to finish its post-processing work - probably not a lot of time individually but it adds up a lot over large runs. I guess that means 2 separate versions, ugh.)

Anyway - sorry this hasn't worked immediately, thanks!

**Comment by thk123** [comment]
Sorry I should have been clearer - the file that worked for me (of the two you originally sent) was the one that started `ntdmc` rather than `ihme` - do you have that one to hand? 

What I've understood from point 2 is - two function calls:

1. Given input directory, aggregate stochastical runs for single IU into standardised format (this function could be run in parallel, or itself could run in parallel with a param for number of jobs, whichever is easier for you)
2. Given a input directory (that should be the output directory of step 1) produce the three aggregate files (probably this has to be singled threaded but I will check if any obviously paralleisable bit).

**Comment by thk123** [comment]
@igorclark I've pushed  WIP parallel solution, not fully tested (but seems to work), though I had to comment out the progress bar for now. By default it uses all but one processor.

**Comment by igorclark** [comment]
Ah wrong files! OK great, I'll try those later today, thanks a lot. And will also try out parallelising the split function. Thanks Thomas 👍

**Comment by thk123** [comment]
@igorclark Just a thought - this script is almost entirely bottlenecked on file IO - I wonder if parallelsing is actually just going to make it slower? Given that the right spreadsheets are much smaller (kb rather than mb) it probably won't make much difference.

**Comment by igorclark** [comment]
Hiya, I tried the latest version using only the `ntdmc` files but got the following error:

```
Post-processing Scenarios: 0it [00:00, ?it/s]
Processing files: 0it [00:00, ?it/s]
Traceback (most recent call last):
  File "/home/igor/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 147, in <module>
    run_postprocessing_pipeline(input_dir="/home/igor/lf-20240830/Outputs/202408a/lf-ntdmc-csv-sample", output_dir=output_dir, num_jobs=2)
  File "/home/igor/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 140, in run_postprocessing_pipeline
    generate_aggregates_for_all_ius(output_iu_dir, output_aggregate_dir)
  File "/home/igor/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 86, in generate_aggregates_for_all_ius
    aggregated_df = iu_lvl_aggregate(combined_ius)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/endgame_postprocessing/post_processing/aggregation.py", line 81, in iu_lvl_aggregate
    aggregated_data.loc[:, columns_to_replace_with_nan] = aggregated_data.loc[
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1184, in __getitem__
    return self._getitem_tuple(key)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1377, in _getitem_tuple
    return self._getitem_tuple_same_dim(tup)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1020, in _getitem_tuple_same_dim
    retval = getattr(retval, self.name)._getitem_axis(key, axis=i)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1420, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1360, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexing.py", line 1558, in _get_listlike_indexer
    keyarr, indexer = ax._get_indexer_strict(key, axis_name)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6200, in _get_indexer_strict
    self._raise_if_missing(keyarr, indexer, axis_name)
  File "/home/igor/.local/share/virtualenvs/endgame-postprocessing-JYDhfSX5/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6249, in _raise_if_missing
    raise KeyError(f"None of [{key}] are in the [{axis_name}]")
KeyError: "None of [Index(['mean'], dtype='object')] are in the [columns]"
```

Here's a [zip](https://github.com/user-attachments/files/16922274/lf-ntdmc-csv-sample.zip) of the first 20 `ntdmc` files in the result list that I'm using, in case I'm still doing something wrong or looking at the wrong thing!

Cheers

**Comment by thk123** [comment]
The above error was caused (I believe) by pointing at the wrong directory, so I have done a bit of work on making the error much clearer. It will now produce the following:

```
/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/post_processing/file_util.py:63: UserWarning: 7 unexpected subdirectories in IU directory lf-ntdmc-csv-sample/lf-ntdmc-csv-sample/lf-ntdmc-csv/lf/scenario_0/SSD, contents will be ignored
  warnings.warn(
/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/post_processing/file_util.py:68: UserWarning: No IU data files found for IU lf-ntdmc-csv-sample/lf-ntdmc-csv-sample/lf-ntdmc-csv/lf/scenario_0/SSD
  warnings.warn(f"No IU data files found for IU {path}")
/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/post_processing/file_util.py:63: UserWarning: 13 unexpected subdirectories in IU directory lf-ntdmc-csv-sample/lf-ntdmc-csv-sample/lf-ntdmc-csv/lf/scenario_0/GNQ, contents will be ignored
  warnings.warn(
/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/post_processing/file_util.py:83: UserWarning: Unexpected file foo.txt in IUs directory lf-ntdmc-csv-sample/lf-ntdmc-csv-sample/lf-ntdmc-csv/lf/scenario_0/GNQ, expecting .csv only
  warnings.warn(
Traceback (most recent call last):
  File "/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 152, in <module>
    run_postprocessing_pipeline(
  File "/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 143, in run_postprocessing_pipeline
    run_aggregate_runs_for_each_iu(input_dir, output_iu_dir, num_jobs=num_jobs)
  File "/home/thomas/dev/endgame-postprocessing/endgame_postprocessing/model_wrappers/lf/testRun.py", line 48, in run_aggregate_runs_for_each_iu
    raise Exception(
Exception: No data for IUs found - see above warnings and check input directory
```

It is still a bit tricky to parse but hopefully is enough for now. But feedback welcome, @igorclark in particular, would the above have been enough do you think? Or would your eyes have glazed over?!

**Comment by igorclark** [comment]
Ha, yeah I think that's fine, the last exception is the key one and the message is clear, nice one!

File: endgame_postprocessing/model_wrappers/lf/testRun.py [hunk 1]
```diff
@@ -12,57 +18,147 @@
...
+        iu_lvl_data=aggregated_df,
+        general_summary_measure_names=[
+            "sampled mf prevalence (all pop)",
+            "year_of_threshold_prevalence_avg",
+            "year_of_90_under_threshold",
+        ],
+        general_groupby_cols=constants.COUNTRY_SUMMARY_GROUP_COLUMNS,
+        threshold_summary_measure_names=[
+            "year_of_threshold_prevalence_avg",
+            "year_of_90_under_threshold",
```
Comment (adiramani) [inline endgame_postprocessing/model_wrappers/lf/testRun.py:111]: Some of these would have to be changed after this PR: https://github.com/NTD-Modelling-Consortium/endgame-postprocessing/pull/8, but I can edit that after this change goes through

**Review by adiramani — APPROVED** [review]
Overall this looks good (other than test cov, which is just a nit). I'm giving my approval, however my only comment would be to generalize the parallel run functionality (if you have time) so that other diseases would be able to leverage it? I think this parallel run process would be useful for oncho if we are post-processing after the fact.

**Comment by thk123** [comment]
Fixed coverage by adding a snapshot test (I think we should have one of these for each model - @adiramani if you could have a look at adding one for oncho?). 

Won't pull out the parallel stuff, can leave for a follow up PR if it makes sense, but should do some quick profiling to see if it is even any faster, as I suspect not.

**Pull request merged**